File: app/src/main/java/ru/startandroid/develop/retrofittask/MessageRepository.java
package ru.startandroid.develop.retrofittask;

import android.app.Application;
import android.os.AsyncTask;

import androidx.lifecycle.LiveData;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class MessageRepository {
    private JsonPlaceHolderApi jsonPlaceHolderApi;
    private MessageDao messageDao;
    private LiveData<List<MessageEntity>> messageEntity;
    private Application application;

    public MessageRepository(Application application) {
        this.application = application;
        MessageDatabase messageDatabase = MessageDatabase.getMessageDatabase(application);
        messageDao = messageDatabase.getMessageDao();
        getMessagesFromServer();
        messageEntity = messageDao.getAllMessage();
    }

    //getMessages from server
    public void getMessagesFromServer() {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://jsonplaceholder.typicode.com/")
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        jsonPlaceHolderApi = retrofit.create(JsonPlaceHolderApi.class);
        Call<List<MessageApiModel>> call = jsonPlaceHolderApi.getMessages();

        call.enqueue(new Callback<List<MessageApiModel>>() {
            @Override
            public void onResponse(Call<List<MessageApiModel>> call, Response<List<MessageApiModel>> response) {
                if (!response.isSuccessful()) {

                    return; //виходить з методу, тут зупиняє виконання методу онРеспонс
                    //брейком можна вийти з циклу або свіча
                }
                //якщо респонс успішний
                List<MessageApiModel> responseFromServer = response.body();
                //додатковий метод який конвертує лісти
                convertToEntity(responseFromServer);
                // presentDataFromDB();
            }

            @Override
            public void onFailure(Call<List<MessageApiModel>> call, Throwable t) {

            }
        });
    }

    public void convertToEntity(List<MessageApiModel> responseFromServer) {

        for (MessageApiModel messageApiModel : responseFromServer) {
            MessageEntity messageEntity = new MessageEntity();
            messageEntity.setId(messageApiModel.getId());
            messageEntity.setUserId(messageApiModel.getUserId());
            messageEntity.setTitle(messageApiModel.getTitle());
            messageEntity.setText(messageApiModel.getText());

            insertMessage(messageEntity);
        }
    }

    public LiveData<List<MessageEntity>> getMessage(){
        return messageEntity;
    }

    public void insertMessage(MessageEntity messageEntity) {
        new InsertMessageAsyncTask(messageDao).execute(messageEntity);
    }

    private static class InsertMessageAsyncTask extends AsyncTask<MessageEntity, Void, Void> {
        private MessageDao messageDao;

        public InsertMessageAsyncTask(MessageDao messageDao) {
            this.messageDao = messageDao;
        }

        @Override
        protected Void doInBackground(MessageEntity... messageEntities) {
            messageDao.insert(messageEntities[0]);
            return null;
        }
    }
}

File: app/src/main/java/ru/startandroid/develop/retrofittask/MessageRepository.kt
package ru.startandroid.develop.retrofittask

import ru.startandroid.develop.retrofittask.MessageDatabase.Companion.getMessageDatabase
import android.app.Application
import androidx.lifecycle.LiveData
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MessageRepository (
        private val application: Application) {
    private lateinit var jsonPlaceHolderApi: JsonPlaceHolderApi
    private val messageDao: MessageDao?
    val message: LiveData<List<MessageEntity>>

    suspend fun fetchData(): List<MessageApiModel>{
        var messageApiModelList: List<MessageApiModel> = emptyList<MessageApiModel>()
        val retrofit = Retrofit.Builder()
                .baseUrl("http://jsonplaceholder.typicode.com/")
                .addConverterFactory(GsonConverterFactory.create())
                .build()
        jsonPlaceHolderApi = retrofit.create(JsonPlaceHolderApi::class.java)

        val call = jsonPlaceHolderApi.messages
        call.enqueue(object : Callback<List<MessageApiModel>> {
            override fun onResponse(call: Call<List<MessageApiModel>>, response: Response<List<MessageApiModel>>) {
                val messages: List<MessageApiModel> = response.body()!!
                messageApiModelList = messages
            }

            override fun onFailure(call: Call<List<MessageApiModel>>, t: Throwable) {}
        })
        return messageApiModelList
    }

    suspend fun fetchAndInsertData(){
        val fetchApiModels = fetchData()
        val convertedList:List<MessageEntity> = convertToEntity(fetchApiModels)
        messageDao!!.insert(convertedList)
    }

    fun convertToEntity(responseFromServer: List<MessageApiModel>) : List<MessageEntity> {
        val list: MutableList<MessageEntity> = ArrayList()
        for (messageApiModel in responseFromServer) {
            val messageEntity = MessageEntity(
            id = messageApiModel.id,
            userId = messageApiModel.userId,
            title = messageApiModel.title,
            text = messageApiModel.text)
            list.add(messageEntity)
        }
        return list
    }

    /*fun insertMessage(messageEntities: List<MessageEntity>) {
        val messageEntityArray = messageEntities.toTypedArray()
        InsertMessageAsyncTask(messageDao).execute(*messageEntityArray)
    }

    private open class InsertMessageAsyncTask(private val messageDao: MessageDao?) : AsyncTask<MessageEntity, Void?, Void?>() {

        override fun doInBackground(vararg messageEntities: MessageEntity): Void? {
            messageDao!!.insert(listOf(*messageEntities))
            return null
        }
    }*/



    init {
        val messageDatabase = getMessageDatabase(application)
        messageDao = messageDatabase.messageDao
        message = messageDao!!.allMessage()
    }
}